pkg/coscheduling/coscheduler.go

```diff
@@ -0,0 +1,484 @@
+	if err != nil {
+		return nil, framework.NewStatus(framework.Unschedulable, err.Error())
+	}
+	if len(node) == 0 {
+		// return skip to make it possible to run default postFilter: preemption
+		return nil, framework.NewStatus(framework.Skip, "")
+	}
+	return &framework.PostFilterResult{NominatedNodeName: node}, framework.NewStatus(framework.Success, "")
+}
+
+// Permit is called before binding a pod (and before prebind plugins). Permit
+// plugins are used to prevent or delay the binding of a Pod. A permit plugin
+// must return success or wait with timeout duration, or the pod will be rejected.
+// The pod will also be rejected if the wait timeout or the pod is rejected while
+// waiting. Note that if the plugin returns "wait", the framework will wait only
+// after running the remaining plugins given that no other plugin rejects the pod.
+func (cs *coSchedulingPlugin) Permit(ctx context.Context, _ *framework.CycleState, p *corev1.Pod, nodeName string) (*framework.Status, time.Duration) {
+	var pgs *schecache.PodGroupSchedulingStatus
+	ready, fullName, err := cs.operation.Permit(ctx, p, nodeName)
+	if fullName != "" {
+		pgObj := cs.operation.PodGroupStatusCache().Get(fullName)
+		if pgObj != nil {
+			pgs = pgObj.(*schecache.PodGroupSchedulingStatus)
+		}
+	}
+	waitTime := util.DefaultWaitTime
+	if pgs != nil {
+		if wait := util.GetWaitTimeDuration(pgs.PodGroup, cs.maxScheduleTime); wait != 0 {
+			waitTime = wait
+		}
+	}
+	// add 1 second to keep the ttl cache would
+	// expired before the the waiting deadline
+	if err != nil {
+		if err == util.ErrorWaiting {
+			klog.Infof("Pod: %v/%v is waiting to be scheduled to node: %v", p.Namespace, p.Name, nodeName)
+			return framework.NewStatus(framework.Wait, ""), waitTime
+		}
+		// For pod not belongs to any groups
+		if err == util.ErrorNotMatched {
+			return framework.NewStatus(framework.Success, ""), 0
+
+		}
+		klog.Infof("cs.operation.Permit error %v", err)
+		return framework.NewStatus(framework.Unschedulable, err.Error()), waitTime
+	}
+	klog.V(5).Infof("Pod requires pgName %v", fullName)
+	if ready {
+		go cs.sendStartScheduleSignal(fullName)
+	}
+
+	return framework.NewStatus(framework.Wait, ""), waitTime
+}
+
+func (cs *coSchedulingPlugin) Reserve(ctx context.Context, state *framework.CycleState, p *corev1.Pod, nodeName string) *framework.Status {
+	return nil
+}
+
+// Unreserve rejects all other Pods in the PodGroup when one of the pods in the group times out.
+func (cs *coSchedulingPlugin) Unreserve(ctx context.Context, state *framework.CycleState, p *corev1.Pod, nodeName string) {
+	pgName, satisfied := util.VerifyPodLabelSatisfied(p)
+	if !satisfied {
+		return
+	}
+	if satisfied && len(pgName) != 0 {
+		fullName := fmt.Sprintf("%v/%v", p.Namespace, pgName)
+		pgsObj := cs.operation.PodGroupStatusCache().Get(fullName)
+		if pgsObj == nil {
+			return
+		}
+		pgs := pgsObj.(*schecache.PodGroupSchedulingStatus)
+		pgs.MatchedPodNodes.Delete(string(p.UID))
```

Comment:
Using local cache, we could avoid frequently iterator the waitPodsMap to choose the pod belong to a group. 
This would also work when we computer resource.

Reply:
Then to keep the cache consistent, upon `Unreserve`, should we also reject the pods in `waitingPodsMap`?